File: test/csharp/SpecGate.CtscFixtures/Conformance/Stateful/SharedSetup/SharedSetup.cs
using SpecGate.Annotations;

namespace SpecGate.CtscFixtures.Conformance.Stateful.SharedSetup;

[SpecEvent("BoxVal")]
public sealed class BoxVal
{
    [SpecEvent("value")]
    public int Value { get; set; }
}

public static class Combine
{
    [SpecSetup("combine", Fills = "left", Spec = "fixture.shared_setup")]
    [SpecSetup("combine", Fills = "right", Spec = "fixture.shared_setup")]
    public static BoxVal MakeBox() => new() { Value = 2 };

    [SpecOperation("combine", Spec = "fixture.shared_setup")]
    public static int Two(
        [SpecInput("left")] BoxVal left,
        [SpecInput("right")] BoxVal right) =>
        left.Value + right.Value;

    [SpecSetup("combine_three", Fills = "a", Spec = "fixture.shared_setup")]
    [SpecSetup("combine_three", Fills = "b", Spec = "fixture.shared_setup")]
    [SpecSetup("combine_three", Fills = "c", Spec = "fixture.shared_setup")]
    public static BoxVal MakeUnit() => new() { Value = 1 };

    [SpecOperation("combine_three", Spec = "fixture.shared_setup")]
    public static int Three(
        [SpecInput("a")] BoxVal a,
        [SpecInput("b")] BoxVal b,
        [SpecInput("c")] BoxVal c) =>
        a.Value + b.Value + c.Value;
}

File: test/rust/crates/specgate-ctsc-fixtures/src/conformance/stateful/shared_setup.rs
use specgate::{SpecEvent, spec_operation, spec_setup};

#[derive(Debug, PartialEq, Eq, SpecEvent)]
#[spec_component("fixture.shared_setup")]
pub struct BoxVal {
    #[spec_event]
    pub value: i32,
}

#[spec_setup("combine", fills = "left", spec = "fixture.shared_setup")]
#[spec_setup("combine", fills = "right", spec = "fixture.shared_setup")]
pub fn make_box() -> BoxVal {
    BoxVal { value: 2 }
}

#[spec_operation("combine", spec = "fixture.shared_setup")]
pub fn combine(left: &BoxVal, right: &BoxVal) -> i32 {
    left.value + right.value
}

#[spec_setup("combine_three", fills = "a", spec = "fixture.shared_setup")]
#[spec_setup("combine_three", fills = "b", spec = "fixture.shared_setup")]
#[spec_setup("combine_three", fills = "c", spec = "fixture.shared_setup")]
pub fn make_unit() -> BoxVal {
    BoxVal { value: 1 }
}

#[spec_operation("combine_three", spec = "fixture.shared_setup")]
pub fn combine_three(a: &BoxVal, b: &BoxVal, c: &BoxVal) -> i32 {
    a.value + b.value + c.value
}

#[test]
fn one_setup_function_can_fill_multiple_parameters() {
    let left = make_box();
    let right = make_box();
    assert_eq!(combine(&left, &right), 4);
    let a = make_unit();
    let b = make_unit();
    let c = make_unit();
    assert_eq!(combine_three(&a, &b, &c), 3);
}
